src/service.py

```diff
@@ -49,24 +51,71 @@ def configure(self, snap_config: dict[str, Any]) -> None:
         self.snap_client.set(snap_config, typed=True)
 
 
-def snap_install(resource: str) -> SnapService:
-    """Install or refresh the snap, and return the snap service.
+def snap_install(resource: Optional[str], channel: str) -> SnapService:
+    """Install the snap, and return the snap service.
+
+    Before installing the snap, it will try to remove the upstream snap that could be installed on
+    older versions of this charm.
+
+    If the channel is changed, the snap.add method is able to identify and refresh the charm to the
+    new channel.
 
     Raises an exception on error.
     """
+    remove_upstream_snap()
     try:
-        logger.debug("installing snap.")
-        logger.debug("fetching from resource.")
-        snap_client = snap.install_local(resource, dangerous=True)
+        if resource:
+            logger.debug("installing %s from resource.", SNAP_NAME)
+            # installing from a resource if installed from snap store previously is not problematic
+            snap_client = snap.install_local(resource, dangerous=True)
+        else:
+            # installing from snap store if previously installed from resource is problematic, so
+            # it's necessary to remove it first
+            remove_snap_as_resource()
+            logger.debug("installing %s from snapcraft store", SNAP_NAME)
+            snap_client = snap.add(SNAP_NAME, channel=channel)
     except snap.SnapError as e:
         logger.error("failed to install snap: %s", str(e))
-        raise e  # need to crash on_install event if it's not okay
+        raise e
     else:
         logger.info("installed %s snap.", snap_client.name)
         workaround_bug_268()
         return SnapService(snap_client)
 
 
+def remove_upstream_snap() -> None:
+    """Remove the old snap from upstream to not conflict with the charmed-openstack-exporter.
+
+    Raises an exception on error.
+    """
+    try:
+        snap.remove(["golang-openstack-exporter"])
```

Comment:
does this function raise an error if golang-openstack-exporter is not present?

Reply:
No, the lib is smart in that sense. It's just an extra cautious if an unknown exception raises